### PR TITLE
docs: forwardref usage for standalone components

### LINKED ISSUE #48614: Cannot import and use parent standalone component in child standalone component
### Which @angular/* package(s) are the source of the bug?

compiler, core

### Is this a regression?

Yes

### Description

I converting my angular application into standalone components, When I have to import parent component in child component then I getting an error. It is works fine with module. In below link I'm importing parent Component in child component after that the app stops working

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-nested-parent-child

### Please provide the exception or error you saw

```true
preview-d0bbfde89eb2a.js:2 ERROR TypeError: Cannot read properties of undefined (reading 'ɵcmp')
    at getComponentDef (container.ts:20:23)
    at extractDirectiveDef (definition.ts:420:6)
    at Array.map (<anonymous>)
    at def.directiveDefs (definition.ts:360:4)
    at createTView (shared.ts:663:4)
    at getOrCreateComponentTView (shared.ts:608:16)
    at addComponentLogic (shared.ts:1422:4)
    at instantiateAllDirectives (shared.ts:1190:4)
    at createDirectivesInstances (shared.ts:561:4)
    at ɵɵelementStart (element.ts:149:6)
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
@angular/animations
15.0.2

@angular/common
15.0.2

@angular/compiler
15.0.2

@angular/core
15.0.2

@angular/forms
15.0.2

@angular/platform-browser
15.0.2

@angular/platform-browser-dynamic
15.0.2

@angular/router
15.0.2

rxjs
7.5.7

tslib
2.4.1

zone.js
0.12.0
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by JeanMeche**
Fixes #48614 and #49059

### CODE REVIEW

File: aio/content/guide/standalone-components.md [hunk 1]
```diff
@@ -341,3 +341,41 @@ In the above example, the component `DateModalComponent` is standalone - it can
...
 
 A separate standalone injector is created to ensure that providers imported by a standalone component are “isolated” from the rest of the application. This lets us think of standalone components as truly self-contained pieces that can’t “leak” their implementation details to the rest of the application.
+
+#### Break circularities with a forward class reference
+
+The order of class declaration matters in TypeScript. You can't refer directly to a class until it's been defined.
+
+This isn't usually a problem but sometimes circular references are unavoidable. You're in a bind when class 'A' refers to class 'B' and 'B' refers to 'A'. One of them has to be defined first.
```
Comment (eneajaho) [inline aio/content/guide/standalone-components.md:349]:
```suggestion
This isn't usually a problem but sometimes circular references are unavoidable. You're in a problematic situation when class 'A' refers to class 'B' and 'B' refers to 'A'. One of them has to be defined first.
```

File: aio/content/guide/standalone-components.md [hunk 2]
```diff
@@ -341,3 +341,41 @@ In the above example, the component `DateModalComponent` is standalone - it can
...
+
+```ts
+@Component({
+  standalone: true, 
+  imports: [ChildComponent],
+  selector: 'app-parent',
+  template: `<app-child [hideParent]="hideParent"></app-child>`,
+})
+export class ParentComponent {
+  @Input() hidParent: boolean;
```
Comment (eneajaho) [inline aio/content/guide/standalone-components.md:364]:
```suggestion
  @Input() hideParent: boolean;
```

File: aio/content/guide/standalone-components.md [hunk 3]
```diff
@@ -341,3 +341,41 @@ In the above example, the component `DateModalComponent` is standalone - it can
...
+
+
+@Component({
+  standalone: true,
+  imports: [CommonModule, forwardRef(() => ParentComponent)],
+  selector: 'app-child',
+  template: `<app-parent *ngIf="!hideParent"></app-parent>`,
+})
+export class ChildComponent {
+  @Input() hidParent: boolean;
```
Comment (eneajaho) [inline aio/content/guide/standalone-components.md:375]:
```suggestion
  @Input() hideParent: boolean;
```

File: packages/core/src/di/forward_ref.ts [hunk 4]
```diff
@@ -34,9 +34,36 @@ const __forward_ref__ = getClosureSafeProperty({__forward_ref__: getClosureSafeP
...
+ * ### Circular standalone reference import example
+ * ```ts
+ * @Component({
+ *   standalone: true,
+ *   imports: [ChildComponent],
+ *   selector: 'app-parent',
+ *   template: `<app-child [hideParent]="hideParent"></app-child>`,
+ * })
+ * export class ParentComponent {
+ *   @Input() hidParent: boolean;
```
Comment (eneajaho) [inline packages/core/src/di/forward_ref.ts:52]:
```suggestion
 *   @Input() hideParent: boolean;
```

File: aio/content/guide/dependency-injection-in-action.md [hunk 5]
```diff
@@ -491,7 +491,7 @@ You can't refer directly to a class until it's been defined.
 
 This isn't usually a problem, especially if you adhere to the recommended *one class per file* rule.
 But sometimes circular references are unavoidable.
-You're in a bind when class 'A' refers to class 'B' and 'B' refers to 'A'.
+You're in a problematic situation when class 'A' refers to class 'B' and 'B' refers to 'A'.
```
Comment (AndrewKushnir) [inline aio/content/guide/dependency-injection-in-action.md:494]: nit: we make it sound a bit more positive :)
```suggestion
For example, when class 'A' refers to class 'B' and 'B' refers to 'A'.
```

File: aio/content/guide/standalone-components.md [hunk 6]
```diff
@@ -341,3 +341,41 @@ In the above example, the component `DateModalComponent` is standalone - it can
 When Angular creates a standalone component, it needs to know that the current injector has all of the necessary services for the standalone component's dependencies, including those based on NgModules. To guarantee that, in some cases Angular will create a new "standalone injector" as a child of the current environment injector. Today, this happens for all bootstrapped standalone components: it will be a child of the root environment injector. The same rule applies to the dynamically created (for example, by the router or the `ViewContainerRef` API) standalone components. 
 
 A separate standalone injector is created to ensure that providers imported by a standalone component are “isolated” from the rest of the application. This lets us think of standalone components as truly self-contained pieces that can’t “leak” their implementation details to the rest of the application.
+
+#### Break circularities with a forward class reference
```
Comment (AndrewKushnir) [inline aio/content/guide/standalone-components.md:345]:
```suggestion
#### Resolve circular dependencies with a forward class reference
```

File: aio/content/guide/standalone-components.md [hunk 7]
```diff
@@ -341,3 +341,41 @@ In the above example, the component `DateModalComponent` is standalone - it can
...
+
+The order of class declaration matters in TypeScript. You can't refer directly to a class until it's been defined.
+
+This isn't usually a problem but sometimes circular references are unavoidable. You're in a problematic situation when class 'A' refers to class 'B' and 'B' refers to 'A'. One of them has to be defined first.
```
Comment (AndrewKushnir) [inline aio/content/guide/standalone-components.md:349]:
```suggestion
This isn't usually a problem but sometimes circular references are unavoidable. For example, when class 'A' refers to class 'B' and 'B' refers to 'A'. One of them has to be defined first.
```

File: aio/content/guide/standalone-components.md (hunk 6 above)
Comment (AndrewKushnir) [reply aio/content/guide/standalone-components.md:345]: I'd propose to update the section in another guide in a similar way.

File: aio/content/guide/standalone-components.md [hunk 8]
```diff
@@ -341,3 +341,41 @@ In the above example, the component `DateModalComponent` is standalone - it can
...
+
+The Angular `forwardRef()` function creates an indirect reference that Angular can resolve later. 
+
+You face this dilemma when a standalone parent component imports a standalone child component and vice-versa. 
+Break the circularity with `forwardRef`.
```
Comment (AndrewKushnir) [inline aio/content/guide/standalone-components.md:354]:
```suggestion
For example, this situation happens when a standalone parent component imports a standalone child component and vice-versa. You can resolve this circular dependency issue by using the `forwardRef` function.
```

File: aio/content/guide/standalone-components.md [hunk 9]
```diff
@@ -341,3 +341,41 @@ In the above example, the component `DateModalComponent` is standalone - it can
...
+  selector: 'app-child',
+  template: `<app-parent *ngIf="!hideParent"></app-parent>`,
+})
+export class ChildComponent {
+  @Input() hideParent: boolean;
+}
+```
+
+<div class="alert is-important">
+This kind of imports may result in an infinite recursion of component instantiation. Make sure to use a condition to control the recursivity.
```
Comment (AndrewKushnir) [inline aio/content/guide/standalone-components.md:380]:
```suggestion
This kind of imports may result in an infinite recursion during component instantiation. Make sure that this recursion has an exit condition that stops it at some point.
```

**Pull request closed without merging**